core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -151,21 +155,49 @@ private synchronized void closeWithTragicEvent(Exception exception) throws IOExc
     /**
      * add the given bytes to the translog and return the location they were written at
      */
-    public synchronized Translog.Location add(BytesReference data) throws IOException {
+
+    /**
+     * Add the given bytes to the translog with the specified sequence number; returns the location the bytes were written to.
+     *
+     * @param data  the bytes to write
+     * @param seqNo the sequence number associated with the operation
+     * @return the location the bytes were written to
+     * @throws IOException if writing to the translog resulted in an I/O exception
+     */
+    public synchronized Translog.Location add(final BytesReference data, final long seqNo) throws IOException {
         ensureOpen();
         final long offset = totalOffset;
         try {
             data.writeTo(outputStream);
-        } catch (Exception ex) {
+        } catch (final Exception ex) {
             try {
                 closeWithTragicEvent(ex);
-            } catch (Exception inner) {
+            } catch (final Exception inner) {
                 ex.addSuppressed(inner);
             }
             throw ex;
         }
         totalOffset += data.length();
         operationCounter++;
+
+        if (minSeqNo == Translog.INITIAL_MIN_SEQ_NO) {
```

Comment:
I'd be happy with `Long.MAX_VALUE` here so long as we're careful not to let them escape from this class.